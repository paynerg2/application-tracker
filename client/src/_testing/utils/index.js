import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../../_reducers';
import { middlewares } from '../../app/store';

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};
