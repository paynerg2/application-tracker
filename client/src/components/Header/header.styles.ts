import { motion } from 'framer-motion';
import styled from 'styled-components';

export const Layout = styled.header`
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 10vh;
    background-color: ${(props) => props.theme.color.background};
    margin-bottom: 2vh;

    @media (max-width: ${(props) => props.theme.breakpoint.laptop}) {
        width: 100vw;
        margin-bottom: 0;
        position: fixed;
        bottom: 0;
        z-index: 100;
        background-color: ${(props) => props.theme.color.footer};
    }
`;

export const HeaderContainer = styled.div`
    width: 80vw;
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;

    @media (max-width: ${(props) => props.theme.breakpoint.laptop}) {
        width: 100%;
        padding-left: 10vw;
        box-sizing: border-box;
    }
`;

export const NavLink = styled.button`
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    font-family: ${(props) => props.theme.font.secondary};
    font-weight: 700;
    color: ${(props) => props.theme.color.lightGray};
    background-color: ${(props) => props.theme.color.background};
    cursor: pointer;
`;

export const SelectedNavLink = styled(NavLink)`
    border-bottom: solid 3px ${(props) => props.theme.color.primary};
    color: ${(props) => props.theme.color.primary};
`;

export const NavLinkSection = styled.nav`
    height: 100%;
    display: flex;
    flex-direction: row;
    gap: 5vw;

    @media (max-width: ${(props) => props.theme.breakpoint.laptop}) {
        display: none;
    }
`;

const itemHeight = '3em';
export const DropdownItem = styled.li`
    width: 100%;
    height: ${itemHeight};
    line-height: ${itemHeight};
    background: ${(props) => props.theme.color.surface};
    display: block;
    transition-duration: 0.2s;
    border-radius: ${(props) => props.theme.borders.radius};
    text-align: center;
    z-index: 1;

    &:hover {
        color: ${(props) => props.theme.color.primary};
        font-weight: 700;
        cursor: pointer;
    }

    &:last-child {
        border-bottom: none;
    }
`;

export const DropdownMenu = styled(motion.ul)`
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 10em;
    height: 6em;
    margin: 0;
    padding-left: 0;
    background-color: ${(props) => props.theme.color.surface};
    border-radius: ${(props) => props.theme.borders.radius};

    position: absolute;
    top: 10vh;
    cursor: pointer;

    // Dialog arrow
    &:before {
        content: '';
        position: absolute;
        width: 2em;
        height: 2em;
        top: -1.2em;
        left: 5.5em;
        background-color: ${(props) => props.theme.color.surface};
        transform: rotate(45deg);
    }
`;

export const IconSection = styled.section`
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    gap: 1vw;
    height: 100%;

    @media (max-width: ${(props) => props.theme.breakpoint.laptop}) {
        display: none;
    }
`;

export const Greeting = styled.div`
    width: min-content;
    color: ${(props) => props.theme.color.mainText};
    // Trick to force one word per line
    word-spacing: 999999px;
`;

export const MobileMenuButton = styled(motion.div)`
    display: none;
    height: 0;

    @media (max-width: ${(props) => props.theme.breakpoint.laptop}) {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 100%;
        max-height: 30em;
        width: 30%;
        background: ${(props) => props.theme.color.button};
        color: ${(props) => props.theme.color.buttonInverted};
        cursor: pointer;
    }
`;
