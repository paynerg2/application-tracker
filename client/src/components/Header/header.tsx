import React, { useRef, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useTheme } from 'styled-components';
import { useAppDispatch } from '../../app/hooks';
import { useAppSelector } from '../../app/hooks';
import useOutsideClick from '../../_helpers/useOutsideClick';
import { api, tags } from '../../services/api';
import { defaultState } from '../../state/authSlice';
import { isEmpty } from '../../_helpers/objectHelpers';
import {
    DropdownItem,
    DropdownMenu,
    Greeting,
    HeaderContainer,
    IconSection,
    Layout,
    NavLink,
    NavLinkSection,
    SelectedNavLink,
    MobileMenuButton,
} from './header.styles';
import ProfileImage from '../ProfileImage/profileImage';
import Logo from '../../assets/Logo.svg';
import DarkLogo from '../../assets/Dark_Logo.svg';
import useWindowDimensions from '../../_helpers/useWindowDimensions';
import { pixelStringToNumber } from '../../_helpers/stringHelpers';
import MobileMenu from '../MobileMenu/mobileMenu';

function Header() {
    const [showMobileMenu, setShowMobileMenu] = useState(false);
    const [showDropdown, setShowDropdown] = useState(false);
    const ref = useRef(null);

    const dispatch = useAppDispatch();
    const location = useLocation();
    const navigate = useNavigate();
    const user = useAppSelector((state) => state.auth.user);
    const { width } = useWindowDimensions();
    const theme = useTheme();

    const pathName = location.pathname.split('/')[1];
    const routes = ['applications', 'interviews', 'stats'];

    useOutsideClick(ref, () => setShowDropdown(false));

    /*-------- Display -------------*/

    const getRouteNav = () => {
        return routes.map((route) =>
            route === pathName.toLocaleLowerCase() ? (
                <SelectedNavLink key={route} onClick={() => handleNavigate(route)}>
                    {route.toLocaleUpperCase()}
                </SelectedNavLink>
            ) : (
                <NavLink key={route} onClick={() => handleNavigate(route)}>
                    {route.toLocaleUpperCase()}
                </NavLink>
            )
        );
    };

    const getLogo = (): string => {
        const breakpoint = pixelStringToNumber(theme.breakpoint.laptop);
        if (width && width <= breakpoint) {
            return DarkLogo;
        }

        return !isEmpty(user) && user.settings?.isDarkMode ? DarkLogo : Logo;
    };

    /*-------- Callbacks -------------*/

    const handleNavigate = (next: string) => {
        navigate(`/${next}`);
    };

    const handleLogout = async () => {
        dispatch(defaultState());
        dispatch(api.util.invalidateTags(tags));
        navigate('/');
    };

    return (
        <Layout>
            <HeaderContainer>
                <img
                    onClick={() => navigate(!isEmpty(user) ? '/applications' : '/')}
                    src={getLogo()}
                    alt="Trackr Logo"
                    style={{ cursor: 'pointer' }}
                />
                <NavLinkSection>{!isEmpty(user) && getRouteNav()}</NavLinkSection>
                {!isEmpty(user) && (
                    <>
                        <IconSection
                            aria-haspopup="true"
                            onMouseEnter={() => setShowDropdown(true)}
                        >
                            {user.fullName && (
                                <Greeting>
                                    Hello, <strong>{user.fullName.split(' ')[0]}</strong>
                                </Greeting>
                            )}
                            <ProfileImage name={user.fullName || ''} />
                            {showDropdown && (
                                <DropdownMenu
                                    ref={ref}
                                    aria-label="submenu"
                                    initial={{ opacity: 0, scale: 0.5 }}
                                    animate={{ opacity: 1, scale: 1 }}
                                    transition={{ duration: 0.2 }}
                                >
                                    <DropdownItem onClick={() => navigate('/me')}>
                                        Edit Profile
                                    </DropdownItem>
                                    <DropdownItem onClick={handleLogout}>Logout</DropdownItem>
                                </DropdownMenu>
                            )}
                        </IconSection>
                        <MobileMenuButton onClick={() => setShowMobileMenu((prev) => !prev)}>
                            {showMobileMenu ? 'X' : 'Menu'}
                        </MobileMenuButton>
                    </>
                )}
            </HeaderContainer>
            <MobileMenu show={showMobileMenu} setShow={setShowMobileMenu} onLogout={handleLogout} />
        </Layout>
    );
}

export default Header;
