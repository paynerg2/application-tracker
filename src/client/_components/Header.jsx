import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

const HeaderContainer = styled.div`
    width: 100%;
    height: 8vh;
    background-color: #bebebe;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`;

const Brand = styled.div`
    font-family: Arial, Helvetica, sans-serif;
    font-size: 30px;
    font-weight: bold;
    padding-left: 29px;
`;

const NavLink = styled(Link)`
    text-decoration: none;
    color: black;
    margin-right: 10px;

    &:focus,
    &:hover,
    &:visited,
    &:link,
    &:active {
        text-decoration: none;
    }
`;

const HeaderLeft = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const HeaderRight = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

export class Header extends Component {
    constructor(props) {
        super(props);

        this.state = {};
    }

    render() {
        return (
            <HeaderContainer>
                <HeaderLeft>
                    <Brand>
                        <NavLink to="/">Application Tracker</NavLink>
                    </Brand>
                    <NavLink to="/data">Data</NavLink>
                    <NavLink to="#">Add New</NavLink>
                </HeaderLeft>
                <HeaderRight>
                    <NavLink to="/login">Login</NavLink>
                    <NavLink to="/register">Sign up</NavLink>
                </HeaderRight>
            </HeaderContainer>
        );
    }
}
