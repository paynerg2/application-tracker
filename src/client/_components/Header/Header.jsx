import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { Link, NavLink } from 'react-router-dom';
import styled from 'styled-components';

import { userActions } from '../../_actions';

const DesktopHeader = styled.div`
    @media (max-width: 760px) {
        display: none;
    }
`;

const MobileHeader = styled.div`
    @media (min-width: 760px) {
        display: none;
    }
`;

const HeaderContainer = styled.div`
    width: 100%;
    height: 10vh;
    background-color: #fff;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    font-family: 'Open Sans', sans-serif;
    font-size: 1.4em;
`;

const MobileHeaderContainer = styled.div`
    width: 100%;
    background-color: #fff;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    font-family: 'Open Sans', sans-serif;
    font-size: 1.4em;
`;

const Brand = styled(Link)`
    color: #1995ad;
    font-family: 'Istok Web', sans-serif;
    font-size: 3em;
    font-weight: bold;
    letter-spacing: 2px;
    padding-left: 29px;
    margin-right: 20px;
    border: none;
    text-decoration: none;

    &:active,
    &:visited,
    &:link,
    &:focus {
        text-decoration: none;
    }

    &:hover {
        text-decoration: none;
        color: #a1d6e2;
    }
`;

const Nav = styled.div`
    text-decoration: none;
    margin-right: 10px;
    padding: 6px;
    color: #1995ad;
`;

const NavigationLink = styled(NavLink)`
    margin: 5px;
    padding: 6px;
    color: #1995ad;

    &:focus,
    &:active,
    &:visited,
    &:hover {
        text-decoration: none;
        color: #1995ad;
    }

    &.is-active {
        text-decoration: none;
        color: white;
        background-color: #1995ad;

        @media (min-width: 760px) {
            color: #1995ad;
            background-color: white;
            border-bottom: solid 3px #1995ad;
        }
    }
`;

const NavButton = styled.button`
    text-decoration: none;
    border: 1px solid #1995ad;
    border-radius: 20px;
    color: #1995ad;
    border: none;
    background-color: transparent;
`;

const MenuIcon = styled.div`
    align-self: flex-end;
    margin-right: 20px;
    color: #666;
    height: 45px;
    padding: 5px;
    font-style: bold;
    border-radius: 6px;
    border: 1px solid #f1f1f2;

    &:hover {
        background-color: #1995ad;
        color: white;
    }
`;

const HeaderLeft = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const HeaderRight = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

class Header extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isExpanded: false
        };

        this.logout = this.logout.bind(this);
    }
    logout() {
        const { dispatch } = this.props;
        dispatch(userActions.logout());
    }

    render() {
        const { user } = this.props;
        return (
            <React.Fragment>
                <DesktopHeader>
                    <HeaderContainer>
                        <HeaderLeft>
                            <Brand to="/">trackr</Brand>
                            <NavigationLink
                                exact
                                to="/"
                                activeClassName="is-active"
                            >
                                Applications
                            </NavigationLink>
                            <NavigationLink
                                exact
                                to="/interviews"
                                activeClassName="is-active"
                            >
                                Interviews
                            </NavigationLink>
                            <NavigationLink
                                exact
                                to="/data"
                                activeClassName="is-active"
                            >
                                Data
                            </NavigationLink>
                            <NavigationLink
                                to="/applications"
                                activeClassName="is-active"
                            >
                                Add New
                            </NavigationLink>
                        </HeaderLeft>
                        <HeaderRight>
                            {user ? (
                                <React.Fragment>
                                    <Nav>{user.username}</Nav>
                                    <Nav>
                                        <NavButton onClick={this.logout}>
                                            Logout
                                        </NavButton>
                                    </Nav>
                                </React.Fragment>
                            ) : (
                                <React.Fragment>
                                    <NavigationLink to="/login">
                                        Login
                                    </NavigationLink>
                                    <NavigationLink to="/register">
                                        Sign up
                                    </NavigationLink>
                                </React.Fragment>
                            )}
                        </HeaderRight>
                    </HeaderContainer>
                </DesktopHeader>
                <MobileHeader>
                    <MobileHeaderContainer>
                        <HeaderContainer>
                            <Brand to="/">trackr</Brand>
                            <MenuIcon
                                onClick={() =>
                                    this.setState({
                                        isExpanded: !this.state.isExpanded
                                    })
                                }
                            >
                                Menu
                            </MenuIcon>
                        </HeaderContainer>
                        {this.state.isExpanded && (
                            <MobileHeaderContainer>
                                <NavigationLink
                                    exact
                                    to="/"
                                    activeClassName="is-active"
                                >
                                    Applications
                                </NavigationLink>
                                <NavigationLink
                                    exact
                                    to="/interviews"
                                    activeClassName="is-active"
                                >
                                    Interviews
                                </NavigationLink>
                                <NavigationLink
                                    exact
                                    to="/data"
                                    activeClassName="is-active"
                                >
                                    Data
                                </NavigationLink>
                                <NavigationLink
                                    to="/applications"
                                    activeClassName="is-active"
                                >
                                    Add New
                                </NavigationLink>
                            </MobileHeaderContainer>
                        )}
                    </MobileHeaderContainer>
                </MobileHeader>
            </React.Fragment>
        );
    }
}

function mapStateToProps(state) {
    const user = state.authentication.user;
    return {
        user
    };
}

Header.propTypes = {
    user: PropTypes.object,
    dispatch: PropTypes.func.isRequired
};

const connectedHeader = connect(mapStateToProps)(Header);
export { connectedHeader as Header };
