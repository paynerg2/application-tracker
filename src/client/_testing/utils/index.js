import checkPropTypes from 'check-prop-types';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../../_reducers';
import { middlewares } from '../../_helpers/store';

export const checkProps = (component, expectedProps) => {
    const propsError = checkPropTypes(
        component.propTypes,
        expectedProps,
        'props',
        component.name
    );
    return propsError;
};

export const testStore = initialState => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(
        createStore
    );
    return createStoreWithMiddleware(rootReducer, initialState);
};
