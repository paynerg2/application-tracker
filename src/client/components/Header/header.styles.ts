import styled from 'styled-components';

export const Layout = styled.header`
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 10vh;
    background-color: ${(props) => props.theme.color.white};
`;

export const HeaderContainer = styled.div`
    width: 80vw;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`;
