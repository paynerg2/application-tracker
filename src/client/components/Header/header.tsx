import React, { useEffect, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

import Logo from '../../assets/Logo.svg';
import { HeaderContainer, Layout, NavLink, NavLinkSection, SelectedNavLink } from './header.styles';

function Header() {
    // TODO: Get user state for user icon/logout, etc.
    const [user, setUser] = useState<string | null>(null);
    const location = useLocation();

    useEffect(() => {
        const user = window.localStorage.getItem('user');
        if (user) {
            setUser(user);
        }
    }, []);

    const routes = ['applications', 'interviews', 'stats'];
    const getRouteNav = () => {
        const pathName = location.pathname.split('/')[1];
        return routes.map((route) =>
            route === pathName.toLocaleLowerCase() ? (
                <SelectedNavLink key={route} to={`/${route}`}>
                    {route.toLocaleUpperCase()}
                </SelectedNavLink>
            ) : (
                <NavLink key={route} to={`${route}`}>
                    {route.toLocaleUpperCase()}
                </NavLink>
            )
        );
    };

    return (
        <Layout>
            <HeaderContainer>
                <Link to="/">
                    <img src={Logo} alt="Trackr Logo" />
                </Link>
                <NavLinkSection>{user && getRouteNav()}</NavLinkSection>
                {user && <div>Logged in as {JSON.parse(user).email!}</div>}
            </HeaderContainer>
        </Layout>
    );
}

export default Header;
