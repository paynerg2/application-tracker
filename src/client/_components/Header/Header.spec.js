import React from 'react';

import { shallow } from 'enzyme';
import { Header } from './Header';
import { testStore } from '../../_testing/utils';

const setUp = (props = {}, initialState = {}) => {
    const store = testStore(initialState);
    const wrapper = shallow(<Header store={store} {...props} />).dive();
    return wrapper;
};

describe('Header Component', () => {
    let wrapper;
    let mockFunction;
    beforeEach(() => {
        mockFunction = jest.fn();
    });

    it('should render without errors', () => {
        wrapper = setUp({ dispatch: mockFunction });
        expect(wrapper).toBeDefined();
    });

    describe('when a user is logged in', () => {
        beforeEach(() => {
            const props = {
                dispatch: mockFunction
            };
            const initialState = {
                authentication: {
                    user: {
                        username: 'test'
                    }
                }
            };
            wrapper = setUp(props, initialState);
        });

        it('should render with user functions', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when a user is not logged in', () => {
        beforeEach(() => {
            const props = {
                dispatch: mockFunction
            };
            wrapper = setUp(props);
        });

        it('should render with authentication functions', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
