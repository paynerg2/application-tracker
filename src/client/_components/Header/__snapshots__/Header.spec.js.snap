// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component when a user is logged in should render with user functions 1`] = `
<styled.div>
  <styled.div>
    <Styled(Link)
      to="/"
    >
      Application Tracker
    </Styled(Link)>
    <Styled(Link)
      to="/interviews"
    >
      Interviews
    </Styled(Link)>
    <Styled(Link)
      to="/data"
    >
      Data
    </Styled(Link)>
    <Styled(Link)
      to="/applications"
    >
      Add New
    </Styled(Link)>
  </styled.div>
  <styled.div>
    <styled.div>
      test
    </styled.div>
    <styled.div>
      <styled.button
        onClick={[Function]}
      >
        Logout
      </styled.button>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Header Component when a user is not logged in should render with authentication functions 1`] = `
<styled.div>
  <styled.div>
    <Styled(Link)
      to="/"
    >
      Application Tracker
    </Styled(Link)>
    <Styled(Link)
      to="/interviews"
    >
      Interviews
    </Styled(Link)>
    <Styled(Link)
      to="/data"
    >
      Data
    </Styled(Link)>
    <Styled(Link)
      to="/applications"
    >
      Add New
    </Styled(Link)>
  </styled.div>
  <styled.div>
    <Styled(Link)
      to="/login"
    >
      Login
    </Styled(Link)>
    <Styled(Link)
      to="/register"
    >
      Sign up
    </Styled(Link)>
  </styled.div>
</styled.div>
`;
